firebase-database/src/main/java/com/google/firebase/database/android/AndroidAuthTokenProvider.java

```diff
@@ -21,65 +21,66 @@
-                    listener.onSuccess(null);
-                  } else {
-                    // TODO: Figure out how to plumb errors through in a sane way.
-                    listener.onError(e.getMessage());
-                  }
-                });
-      }
+    deferredAuthProvider.whenAvailable(authProvider -> internalAuth.set(authProvider.get()));
+  }
+
+  @Override
+  public void getToken(boolean forceRefresh, @NonNull final GetTokenCompletionListener listener) {
+    InternalAuthProvider authProvider = internalAuth.get();
 
-      @Override
-      public void addTokenChangeListener(
-          final ExecutorService executorService, final TokenChangeListener tokenListener) {
-        IdTokenListener idTokenListener =
-            tokenResult ->
-                executorService.execute(
-                    () -> tokenListener.onTokenChange(/* nullable */ tokenResult.getToken()));
-        authProvider.addIdTokenListener(idTokenListener);
-      }
+    if (authProvider != null) {
+      // TODO: This would be easier if Deferred<T> had a "isAvailable()" API
+      Task<GetTokenResult> getTokenResult = authProvider.getAccessToken(forceRefresh);
 
-      @Override
-      public void removeTokenChangeListener(TokenChangeListener tokenListener) {
-        // TODO Implement removeIdTokenListener.
-      }
-    };
+      getTokenResult
+          .addOnSuccessListener(result -> listener.onSuccess(result.getToken()))
+          .addOnFailureListener(
+              e -> {
+                if (isUnauthenticatedUsage(e)) {
+                  listener.onSuccess(null);
+                } else {
+                  // TODO: Figure out how to plumb errors through in a sane way.
+                  listener.onError(e.getMessage());
+                }
+              });
+    } else {
+      listener.onSuccess(null);
+    }
   }
 
-  public static AuthTokenProvider forUnauthenticatedAccess() {
-    return new AuthTokenProvider() {
-      @Override
-      public void getToken(boolean forceRefresh, GetTokenCompletionListener listener) {
-        listener.onSuccess(null);
-      }
+  @Override
+  public void addTokenChangeListener(
+      final ExecutorService executorService, final TokenChangeListener tokenListener) {
+    InternalAuthProvider authProvider = internalAuth.get();
+    if (authProvider == null) {
+      executorService.execute(() -> tokenListener.onTokenChange(null));
+    }
 
-      @Override
-      public void addTokenChangeListener(
-          ExecutorService executorService, TokenChangeListener listener) {
-        executorService.execute(() -> listener.onTokenChange(null));
-      }
+    IdTokenListener idTokenListener =
+        tokenResult ->
+            executorService.execute(
+                () -> tokenListener.onTokenChange(/* nullable */ tokenResult.getToken()));
+    deferredAuthProvider.whenAvailable(
+        // TODO: Could "whenAvailable" return the Auth instance rather than a provider for Auth?
```

Comment:
The return of the provider is intentional since it's more general and makes things lazy by default - so that we don't have to initialize with inside components unless get() is called on its provider.

Reply:
Could `Deferred` be modified to extend `Provider`? As I'm working on changes for Firestore, this would be incredibly convenient because some usages need to react instantly when the provider becomes available (i.e. via `whenAvailable()`) and other instances need to get the instance and do something with it if it is non-null (which would be achieved by calling `get()`). 

Since [OptionalProvider](https://github.com/firebase/firebase-android-sdk/blob/4a366a823000b37790c1af4145366fee914489e3/firebase-components/src/main/java/com/google/firebase/components/OptionalProvider.java#L30) already implements both `Deferred` and `Provider`, this would be an incredibly easy change to implement. Also, it would satisfy @schmidt-sebastian's request for an `isAvailable()` method.

Reply:
@vkryachko I've implemented this suggestion in https://github.com/firebase/firebase-android-sdk/pull/2432. I'll officially add you both as reviewers once the GitHub Actions pass.